### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by sbj42**
This change converts `nodeHash`, `pointsToCost`, `directionalConditions`, and `pointsToAvoid` into two-level hashes.  Instead of accessing them via `hash[x + '_' + y]`, this basically uses `hash[y][x]`.  That improves performance by avoiding string concatenation.

In my tests, that speeds it up between 2x and 6x, depending on the benchmark.

This PR also includes some cleanup in `calculate()`: getting rid of `tilesToSearch` and `isDoneCalculating`.

### CODE REVIEW

File: src/easystar.js [hunk 1]
```diff
@@ -362,27 +374,29 @@ EasyStar.js = function() {
                 path.reverse();
                 var ip = path;
                 instance.callback(ip);
-                return
```
Comment (prettymuchbryce) [inline src/easystar.js]: Nice. This looks like it was actually a bug. The calculation should have continued for other instances here. Nice find.

File: src/easystar.js [hunk 2]
```diff
@@ -422,31 +435,12 @@ EasyStar.js = function() {
...
-
-                        tilesToSearch.push({ instance: instance, searchNode: searchNode,
-                            x: -1, y: 1, cost: DIAGONAL_COST * getTileCost(searchNode.x-1, searchNode.y+1)});
+                        checkAdjacentNode(instance, searchNode,
+                            -1, 1, DIAGONAL_COST * getTileCost(searchNode.x-1, searchNode.y+1));
                     }
                 }
             }
 
-            var isDoneCalculating = false;
```
Comment (prettymuchbryce) [inline src/easystar.js]: This is much cleaner after having removed this, and the loop below. 👍

File: src/easystar.js [hunk 3]
```diff
@@ -510,12 +506,16 @@ EasyStar.js = function() {
...
 
     var getTileCost = function(x, y) {
-        return pointsToCost[x + '_' + y] || costMap[collisionGrid[y][x]]
+        return (pointsToCost[y] && pointsToCost[y][x]) || costMap[collisionGrid[y][x]]
     };
 
     var coordinateToNode = function(instance, x, y, parent, cost) {
-        if (instance.nodeHash[x + "_" + y]!==undefined) {
-            return instance.nodeHash[x + "_" + y];
+        if (instance.nodeHash[y]!==undefined) {
```
Comment (prettymuchbryce) [inline src/easystar.js]: Would you mind adding spaces around `!==`? I need to get around to adding a linter to this codebase.

**Comment by prettymuchbryce** [comment]
I mentioned this in the other PR as well, but could you remove the changes to files in `/bin/`?

**Comment by sbj42** [comment]
Sure no problem.

**Pull request merged**